subsys/modem/modem_pipe.c

```diff
@@ -6,8 +6,74 @@
 
 #include <zephyr/modem/pipe.h>
 
-#include <zephyr/logging/log.h>
-LOG_MODULE_REGISTER(modem_pipe, CONFIG_MODEM_MODULES_LOG_LEVEL);
+#define PIPE_EVENT_OPENED_BIT        BIT(0)
+#define PIPE_EVENT_CLOSED_BIT        BIT(1)
+#define PIPE_EVENT_RECEIVE_READY_BIT BIT(2)
+#define PIPE_EVENT_TRANSMIT_IDLE_BIT BIT(3)
+
+static void pipe_set_callback(struct modem_pipe *pipe,
+			      modem_pipe_api_callback callback,
+			      void *user_data)
+{
+	K_SPINLOCK(&pipe->spinlock) {
+		pipe->callback = callback;
+		pipe->user_data = user_data;
+	}
+}
+
+static void pipe_call_callback(struct modem_pipe *pipe, enum modem_pipe_event event)
+{
+	K_SPINLOCK(&pipe->spinlock) {
+		if (pipe->callback != NULL) {
+			pipe->callback(pipe, event, pipe->user_data);
+		}
+	}
+}
+
+static uint32_t pipe_test_events(struct modem_pipe *pipe, uint32_t events)
+{
+	return k_event_test(&pipe->event, events);
+}
+
+static uint32_t pipe_await_events(struct modem_pipe *pipe, uint32_t events)
+{
+	return k_event_wait(&pipe->event, events, false, K_MSEC(10000));
```

Comment:
Should we have a kconfig or define for this `10000` value?

Reply:
Already on it, I have an upcomming PR which updates the APIs modem_pipe_open() and modem_pipe_close() to take a timeout parameter `modem_pipe_open(struct modem_pipe *pipe, k_timeout_t timeout);` and updates to the in-tree drivers :)